server: make the --insecure warning more explicit.

knz opened the pull request:
Example output:
```
*
* WARNING: RUNNING IN INSECURE MODE!
*
* - Your cluster is open for any client that can access localhost.
* - Any user, even root, can log in without providing a password.
* - Any user, connecting as root, can read or write any data in your cluster.
* - There is no network encryption nor authentication, and thus no confidentiality.
*
* Check out how to secure your cluster: https://www.cockroachlabs.com/docs/secure-a-cluster.html
*
CockroachDB node starting at 2017-05-15 14:07:53.1116454 +0000 UTC
build:      CCL 12ec77f-dirty @ 2017/05/15 14:05:14 (go1.8)
admin:      http://localhost:8080
sql:        postgresql://root@localhost:26257?sslmode=disable
logs:       /home/kena/cockroach/cockroach-data/logs
store[0]:   path=/home/kena/cockroach/cockroach-data
status:     restarted pre-existing node
clusterID:  fe26b6f3-3153-4c75-a7fb-3027c327e5ab
nodeID:     1

```

cc @mberhault

knz committed:
server: make the --insecure warning more explicit.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15724)
<!-- Reviewable:end -->

petermattis commented:
The additional warning commentary is nicely eye catching, but the 10 second delay feels over the top (IMO). And this delay is only performed when stdin is a tty which seems exactly when I don't want a delay.

knz commented:
I'm not sure how long is reasonable. There certainly needs to be some delay to give the user some time to read the message and react. My rationale for checking the TTY is that there's not much for the user to interact if stdout or stdin is redirected. I am open for suggestions.

petermattis commented:
Does there need to be a delay? Running in insecure mode is not the default. The user must have typed `--insecure` explicitly.

knz commented:
What if it's their first time and they haven't yet understood what it was really about? (I certainly hadn't many times before I did, and neither did Jordan)

petermattis commented:
> What if it's their first time and they haven't yet understood what it was really about?

If they are starting a cluster for the first time, is there data that needs protecting?

knz commented:
That's not the only use case. One can create a cluster securely, and only later try to start a node insecurely for the existing cluster.

petermattis commented:
Well, one insecure node won't be able to communicate with the rest of the cluster. I'm going to sleep on this, though. Perhaps a definitive argument will reveal itself to me in the night.

knz commented:
Sleep tight!

a-robinson commented on pkg/cli/cliflags/flags.go:
```diff
@@ -255,8 +255,11 @@ production usage.`,
-Start an insecure node, using unencrypted (non-TLS) connections. This is strongly discouraged for
-production usage.`,
+Start an insecure node, using unencrypted (non-TLS) connections,
+listening on all IP addresses and disabling password authentication
```
"listening on all IP addresses" isn't true - that's controlled by the host flag, as recently changed by Marc's certs overhaul.

a-robinson commented on pkg/cli/cliflags/flags.go:
```diff
@@ -255,8 +255,11 @@ production usage.`,
+listening on all IP addresses and disabling password authentication
+for all database users. This is strongly discouraged for production
+usage and should never be used on a public network without combining
+it with --host and --http-host.`,
```
To be fair, the admin UI doesn't contain much sensitive info. It might be better to say that it should never be used on a public network with data that you care about at all.

a-robinson commented on pkg/server/server.go:
```diff
@@ -159,7 +163,23 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
+				"- your cluster is OPEN for any client that can access "+addr+"!\n"+
+				"- any user even 'root' can log in WITHOUT PROVIDING A PASSWORD!\n"+
+				"- any user can READ or WRITE ANY DATA in your cluster!\n"+
+				"- there is NO NETWORK ENCRYPTION, NO TLS, NO CONFIDENTIALITY!")
```
A call to action here would be helpful for people that see this warning and get scared but don't know what to do about it. Something like "For information about how to secure a cluster, please see the documentation at https://www.cockroachlabs.com/docs/secure-a-cluster.html"

a-robinson commented on pkg/server/server.go:
```diff
@@ -159,7 +163,23 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
+				"- there is NO NETWORK ENCRYPTION, NO TLS, NO CONFIDENTIALITY!")
+		if isInteractive {
+			log.Shout(context.Background(), log.Severity_INFO,
+				"Waiting 10 seconds. Press Ctrl+C now to abort!")
```
I'm also not a big fan of waiting, since as @petermattis said when a cluster is initialized it doesn't have any data to protect, and insecure nodes can't be added to a cluster without the admin having already decided that the cluster being insecure is acceptable.

a-robinson reviewed (commented):
I like the extra detail in the warning to help users make informed decisions, but stylistically think it's a bit over the top. I'm curious what @jseldess thinks.

benesch commented on pkg/server/server.go:
```diff
@@ -159,7 +163,23 @@ func NewServer(cfg Config, stopper *stop.Stopper) (*Server, error) {
+		log.Shout(context.Background(), log.Severity_WARNING,
+			"'INSECURE' IS REALLY INSECURE:\n"+
+				"- your cluster is OPEN for any client that can access "+addr+"!\n"+
+				"- any user even 'root' can log in WITHOUT PROVIDING A PASSWORD!\n"+
```
any user, even 'root', can log in...

(To my ear, this isn't correct without the commas.)

benesch commented:
I'd advocate for these copy-edits, but I don't feel strongly:

```
*
* WARNING: RUNNING IN INSECURE MODE!
*
*
* INSECURE MODE IS REALLY INSECURE:
*     - Your cluster is OPEN to any client that can access <your IP address>:26257!
*     - Any user, even root, can log in WITHOUT PROVIDING A PASSWORD!
*     - Any user can READ or WRITE ANY DATA in your cluster!
*     - There is NO NETWORK ENCRYPTION, NO TLS, and NO CONFIDENTIALITY!
*
*
* INFO: Waiting 10 seconds. Press Ctrl+C now to abort!
*
```

knz commented:
I removed the wait as requested.

---

Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/cli/cliflags/flags.go, line 259 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjQ2H-M495PqGGloUTS-r1-259:-KjT0ft2M21KhrR5Sb-M:b-9zbuyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a9d728ce94ab255185e7f92a0468675255508c0/pkg/cli/cliflags/flags.go#L259)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

"listening on all IP addresses" isn't true - that's controlled by the host flag, as recently changed by Marc's certs overhaul.
</blockquote></details>

The default for --host is empty and that defaults to listening on all interfaces. I added this detail.

---

*[pkg/cli/cliflags/flags.go, line 262 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjQ2H-M495PqGGloUTS-r1-262:-KjT0hMEgxF84wca1k8T:b-222dna) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a9d728ce94ab255185e7f92a0468675255508c0/pkg/cli/cliflags/flags.go#L262)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

To be fair, the admin UI doesn't contain much sensitive info. It might be better to say that it should never be used on a public network with data that you care about at all. 
</blockquote></details>

Not strictly correct: there are probably enough bugs in CockroachDB that an open port == remote attack vector onto the entire system for an attacker.
I restricted the note to --host. I think you're right about the http endpoint, that is probably already restricted enough.

---

*[pkg/server/server.go, line 175 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjQ2H-M495PqGGloUTT-r1-175:-KjT19ueDXVHdkJ596Em:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a9d728ce94ab255185e7f92a0468675255508c0/pkg/server/server.go#L175)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

any user, even 'root', can log in...

(To my ear, this isn't correct without the commas.)
</blockquote></details>

Done.

---

*[pkg/server/server.go, line 177 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjQ2H-M495PqGGloUTT-r1-177:-KjT1LO2NqNIzmGfBQZ1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a9d728ce94ab255185e7f92a0468675255508c0/pkg/server/server.go#L177)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

A call to action here would be helpful for people that see this warning and get scared but don't know what to do about it. Something like "For information about how to secure a cluster, please see the documentation at https://www.cockroachlabs.com/docs/secure-a-cluster.html"
</blockquote></details>

Done.

---

*[pkg/server/server.go, line 180 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjQ2H-M495PqGGloUTT-r1-180:-KjT1MYfdUJtmjasOeoB:b-40kt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a9d728ce94ab255185e7f92a0468675255508c0/pkg/server/server.go#L180)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'm also not a big fan of waiting, since as @petermattis said when a cluster is initialized it doesn't have any data to protect, and insecure nodes can't be added to a cluster without the admin having already decided that the cluster being insecure is acceptable.
</blockquote></details>

Removed the wait.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjT18XuEBjr24JKDXzg:bnorz9q)*
<!-- Sent from Reviewable.io -->

jseldess commented:
It's great that we're being much more explicit, but it goes a bit far for me. I'd say we can remove the second line of caps,  `INSECURE MODE IS REALLY INSECURE`, since the first line is an exclamatory warning. 

Also, is `Any user can READ or WRITE ANY DATA in your cluster!` really true? Don't privileges still control user-specific access to dbs and tables? If so, I think we can remove this bullet and still have a very strong message:

*
* WARNING: RUNNING IN INSECURE MODE!
*
*     - Your cluster is OPEN to any client that can access <your IP address>:26257!
*     - Any user, even root, can log in WITHOUT PROVIDING A PASSWORD!
*     - There is NO NETWORK ENCRYPTION, NO TLS, and NO CONFIDENTIALITY!
*

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjVMurKlX7XhYoPU9X9:b-j8ijvx)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Or is `Any user can READ or WRITE ANY DATA in your cluster` supposed to connect to the fact that any user can log in as `root`? 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjVOqGOt4MhdH4dBrlw:b-espzm9)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'm ok with a louder warning, but it is sort of strange we feel it is necessary. Isn't the `--insecure` flag sufficient? Should we rename the flag to `--insecure-we-really-mean-it`?

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjVS-sMdjfb-X7dh62d:b2llmp5)*
<!-- Sent from Reviewable.io -->

jseldess commented:
I think this bubbled up because people were surprised by the fact that passwords aren't respected in insecure clusters, right? I agree with Pete, though. The rest seems already clear (cluster being open, no encryption, etc.), and so I also find this kind of overboard. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjVUqU-wbLGxx1ZvkRS:b-eajnwm)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
I think the trouble is that `--insecure` is overloaded at the moment, to mean 1. certs are not required and 2. passwords are not required. Most databases don't require setting up SSL certificates to run, but they still respect various password settings. I don't think most databases have a mode that simply ignores all authentication even though it might already have been set up.

Because of this, I think that users intimidated by the SSL certificate set up steps might decide to use `--insecure` as a specific workaround for missing certs. They might be surprised that this "workaround" also completely disables all of the other forms of authentication that the database has.

Therefore I'm in favor of the louder warning. It would have helped me!

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjVXQAg-Lh-gUzmcuBl:buza9t3)*
<!-- Sent from Reviewable.io -->

knz commented:
@jesse: yes it is connected to that fact. Privilege controls are still active, but since any user can connect as root, effectively any user can read/write any data.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-:-KjZQRwNOW2W-BMGRd9O:bn044c8)*
<!-- Sent from Reviewable.io -->

knz commented:
@jesse as you suggested I have shortened the message:
```
*
* WARNING: RUNNING IN INSECURE MODE!
*
* - Your cluster is OPEN for any client that can access <all your IP addresses>!
* - Any user, even root, can log in WITHOUT PROVIDING A PASSWORD!
* - Any user can READ or WRITE ANY DATA in your cluster!
* - There is NO NETWORK ENCRYPTION, NO TLS, and NO CONFIDENTIALITY!
*
* Check out how to secure your cluster: https://www.cockroachlabs.com/docs/secure-a-cluster.html
*
CockroachDB node starting at 2017-05-07 20:11:45.605402227 +0000 UTC
[...]
```

jseldess commented:
Got it, @knz. What would you think about refining the third bullet a bit?

```
* - A user connecting as root can READ or WRITE ANY DATA in your cluster!
```

knz commented:
Done

jseldess commented:
LGTM

knz commented:
@a-robinson gtg?

a-robinson commented on pkg/cli/start.go:
```diff
@@ -569,6 +569,28 @@ func setupAndInitializeLoggingAndProfiling(startCtx context.Context) (*stop.Stop
+		go func() {
+			for range time.Tick(5 * time.Minute) {
+				log.Shout(context.Background(), log.Severity_WARNING,
+					"RUNNING IN INSECURE MODE - THE DATA IN THIS CLUSTER IS UNSAFE")
```
Is shouting this every 5 minutes really necessary? I'd almost think that even just logging it every 5 minutes is unnecessary, but this definitely feels like too much. It'd make running a local cluster with `--background` super annoying.

bdarnell replied:
Yeah, the repeated warning seems really unnecessary. If they didn't see it the first time, repeating it is unlikely to help, and if they did see it then it's just going to be annoying.

knz commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 589 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjTCcdOTYpgEZ7QRuPq-r2-589:-KjxCMpkTnsYz2eZQyMT:b-34g5pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ad0ea6097373d8b4faf8a812a81d33262128be/pkg/cli/start.go#L589)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, the repeated warning seems really unnecessary. If they didn't see it the first time, repeating it is unlikely to help, and if they did see it then it's just going to be annoying. 
</blockquote></details>

maybe shouting is not necessary, but I think a reminder in a log file is very well welcome, because log file rotation may cause the initial warning to disappear (if we receive a log file for troubleshooting, for example)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 589 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjTCcdOTYpgEZ7QRuPq-r2-589:-Kk7RVXgdTlzoPB8Scyo:b8zw8jq) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ad0ea6097373d8b4faf8a812a81d33262128be/pkg/cli/start.go#L589)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

maybe shouting is not necessary, but I think a reminder in a log file is very well welcome, because log file rotation may cause the initial warning to disappear (if we receive a log file for troubleshooting, for example)
</blockquote></details>

The insecure status of a cluster is unlikely to be relevant to our troubleshooting efforts, though, so is the purpose of including it in the logs just to get us to scold the user if they send us a log showing that they're using an insecure cluster in production?

The goal is to make this condition visible to the user/admin, and the logs are a poor place to do that. I'd rather do this as a banner in the admin UI or whenever the sql shell is used than to spam the logs with it. (and of course making it visible to the user that they are making a mistake is a poor substitute for helping them to avoid the mistake in the first place. We need to eliminate `--insecure` from most of our docs and show the secure way from the beginning)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 589 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjTCcdOTYpgEZ7QRuPq-r2-589:-KkBEV7xATx-wBqjbZSs:b-mmteme) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ad0ea6097373d8b4faf8a812a81d33262128be/pkg/cli/start.go#L589)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The insecure status of a cluster is unlikely to be relevant to our troubleshooting efforts, though, so is the purpose of including it in the logs just to get us to scold the user if they send us a log showing that they're using an insecure cluster in production?

The goal is to make this condition visible to the user/admin, and the logs are a poor place to do that. I'd rather do this as a banner in the admin UI or whenever the sql shell is used than to spam the logs with it. (and of course making it visible to the user that they are making a mistake is a poor substitute for helping them to avoid the mistake in the first place. We need to eliminate `--insecure` from most of our docs and show the secure way from the beginning)
</blockquote></details>

I think I wasn't explicit enough. When a user runs an insecure cluster and attracts "unwanted clients" (from script kiddies to plain malicious worms) and start experiencing node crashes or misbehavior due to said unwanted clients, the first reaction will be "why are my nodes crashing" and we'll get the log file with the question "what went wrong". I want to spend a maximum of zero complete minutes investigating this, and be able to say "this was insecure? you deal with it". I'd say it's either a message in the log regularly so we know about it despite the rotation, *or* an extra character on every log line generated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/cli/start.go, line 589 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724#-KjTCcdOTYpgEZ7QRuPq-r2-589:-KkBI5COUOSaYhEJeZok:bva14rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ad0ea6097373d8b4faf8a812a81d33262128be/pkg/cli/start.go#L589)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think I wasn't explicit enough. When a user runs an insecure cluster and attracts "unwanted clients" (from script kiddies to plain malicious worms) and start experiencing node crashes or misbehavior due to said unwanted clients, the first reaction will be "why are my nodes crashing" and we'll get the log file with the question "what went wrong". I want to spend a maximum of zero complete minutes investigating this, and be able to say "this was insecure? you deal with it". I'd say it's either a message in the log regularly so we know about it despite the rotation, *or* an extra character on every log line generated.
</blockquote></details>

Eh, if a node is crashing, we should troubleshoot and fix the crash, "insecure" mode or not.

I'm -1 on the blanket condemnation of running without TLS -- if you're already using trusted network (VPN, trusted LAN, host/vm, etc), you might legitimately want the simpler deployment -- even if we make TLS easy, many client libraries don't.

Seems like the big gotcha here, that we're throwing this aggressive warning at, is that `--insecure` disables password verification: many users might not want and indeed be very surprised by this, and it's conflation with enabling/disabling TLS is why we have and expect them to be using it to control TLS. Maybe we should just disentangle these two options, to eliminate that surprise and confusion at its source?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15724)*
<!-- Sent from Reviewable.io -->

knz commented:
Oh and to react to the 2nd half of your response, yes **absolutely** the proper thing to do here is to have --insecure only remove network encryption and keep all the other security features in place. **however** the pgwire authentication scheme we currently support is insecure AF when there is no network encryption (= completely vulnerable to mitms) so it's not very different from the current situation.

**if/when** we support one of the more secure authentication protocols, we can revisit sane options. Until then, users must be informed and we must disclaim liability at every level of the use-support lifecycle.

dt commented:
> With --insecure, beyond "no encryption" there is no security. No authentication!
@knz Right, that's precisely why I was recommending we fix _that_, not shout at the user about it.

> Privilege checks out of the window. "Debug functions" in our builtins package exposed to any user.
My understanding, is that we still do all those things correctly, and the only issue is that anyone can connect as root, which... see above.

>  pgwire authentication scheme we currently support is insecure AF when there is no network encryption
Which is why we recommend TLS everywhere -- but again, if you're on a trusted *network* and want to skip TLS, it's pretty weird to also skip auth.

mberhault commented:
I've been staying away since this PR was obviously going to devolve into an argument.

However, I'll take a few sides:
* I really don't think we need the recurring warning, even without shouting
* this is overly loud. The first line is already all caps, we don't need to keep shouting at people
* `NO NETWORK ENCRYPTION, NO TLS, and NO CONFIDENTIALITY!` is just three ways of saying the same thing. Simply `There is no network encryption or authentication` should be sufficient.

As for future changes, I agree that we should allow passwords in insecure mode. Even if you accept the risk of MITM attacks, you may want passwords to prevent operator error. But this is outside the scope of this PR, we're not going to do it in 1.0.x

knz commented:
Ah hmm I start to realize there's a thought at the back of my mind which I was assuming was shared but which haven't expressed at all until now. Sorry for that:

My understanding is that the motivation for --insecure is to skip certificate management with an expectation that *inter-node* traffic is safe/confidential, but that users who go with --insecure still have an expectation that client-node traffic is protected properly. 

To address this, we'd need a mode where where clients connect over TLS without client certificate checks from the server (but potentially enabling server certificate verification by the client, so that MITM attacks are foiled), with a proper authentication protocol. That's what "splitting the concerns" means to me.

@mberhault  I'm ok with lowercase messages. I'll see what I can do.

As to the repeat logging, I am not religious about the form but I am still quite keen on finding a way for us to recognize the situation from a mile away and renounce further investigation. Perhaps another way to achieve this is to embed the insecure marker at the beginning of the log file, together with the header that reproduces the node version number. Thoughts?

mberhault commented:
We already do:
```
$ cockroach start --insecure
$ head cockroach-data/logs/cockroach.fishmonger.marc.2017-05-15T09_59_18-04_00.018673.log
I170515 09:59:18.797062 1 util/log/clog.go:1011  [config] file created at: 2017/05/15 09:59:18
I170515 09:59:18.797062 1 util/log/clog.go:1011  [config] running on machine: fishmonger
I170515 09:59:18.797062 1 util/log/clog.go:1011  [config] binary: CockroachDB CCL v1.0 (linux amd64, built 2017/05/15 12:27:49, go1.8.1)
I170515 09:59:18.797062 1 util/log/clog.go:1011  [config] arguments: [cockroach start --insecure]
I170515 09:59:18.797062 1 util/log/clog.go:1011  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170515 09:59:18.797061 1 cli/start.go:575  CockroachDB CCL v1.0 (linux amd64, built 2017/05/15 12:27:49, go1.8.1)
I170515 09:59:18.915009 52 cli/start.go:338  starting cockroach node
W170515 09:59:18.915078 52 server/server.go:162  running in insecure mode, this is strongly discouraged. See --insecure.
I170515 09:59:18.942697 52 storage/engine/rocksdb.go:380  opening rocksdb instance at "/home/marc/cockroach/src/github.com/cockroachdb/cockroach/cockroach-data"
I170515 09:59:19.274206 52 server/config.go:421  1 storage engine initialized
```

This is repeated for every new log file.

As for authentication modes, we can already run the nodes in secure mode and skip client certificates.We just don't document this well and don't support it fully through the flag-based `cockroach sql` command. Using `--url` works though. See: #15318.

knz commented:
Regarding my last suggestion: I can see the arguments are already logged in the header :smile: Excellent, that indeed removes the need for periodic logging.

knz commented:
Here's the new message as suggested by Marc:

```
*
* WARNING: RUNNING IN INSECURE MODE!
*
* - Your cluster is open for any client that can access localhost.
* - Any user, even root, can log in without providing a password.
* - Any user, connecting as root, can read or write any data in your cluster.
* - There is no network encryption nor authentication, and thus no confidentiality.
*
* Check out how to secure your cluster: https://www.cockroachlabs.com/docs/secure-a-cluster.html
*
CockroachDB node starting at 2017-05-15 14:07:53.1116454 +0000 UTC
build:      CCL 12ec77f-dirty @ 2017/05/15 14:05:14 (go1.8)
admin:      http://localhost:8080
sql:        postgresql://root@localhost:26257?sslmode=disable
logs:       /home/kena/cockroach/cockroach-data/logs
store[0]:   path=/home/kena/cockroach/cockroach-data
status:     restarted pre-existing node
clusterID:  fe26b6f3-3153-4c75-a7fb-3027c327e5ab
nodeID:     1
```

mberhault commented:
LGTM

bdarnell commented:
> Oh and to react to the 2nd half of your response, yes absolutely the proper thing to do here is to have --insecure only remove network encryption and keep all the other security features in place. however the pgwire authentication scheme we currently support is insecure AF when there is no network encryption (= completely vulnerable to mitms) so it's not very different from the current situation.

Even if we supported a better mode of authentication for pgwire, the bigger problem is that the inter-node GRPC connections are secured only by TLS certs, allowing for trivial root/node-level access when TLS is disabled. We would need non-TLS authentication for GRPC in addition to better pgwire password authentication to allow `--insecure` to provide even partial security. Until we change the GRPC auth, adding passwords to `--insecure` mode just gives a false sense of security.

knz commented:
Thanks for all the feedback! TFYRs too.
